tools/compose.c

```diff
@@ -108,18 +144,40 @@ main(int argc, char *argv[])
         goto out;
     }
 
-    compose_table =
-        xkb_compose_table_new_from_locale(ctx, locale,
-                                          XKB_COMPOSE_COMPILE_NO_FLAGS);
-    if (!compose_table) {
-        fprintf(stderr, "Couldn't create compose from locale\n");
-        goto out;
+    if (path != NULL) {
+        FILE *file = fopen(path, "rb");
+        if (file == NULL) {
+            perror(path);
+            goto file_error;
+        }
+        compose_table =
+            xkb_compose_table_new_from_file(ctx, file, locale, format,
+                                            XKB_COMPOSE_COMPILE_NO_FLAGS);
+        fclose(file);
+        if (!compose_table) {
+            fprintf(stderr, "Couldn't create compose from file: %s\n", path);
+            goto out;
+        }
+    } else {
+        compose_table =
+            xkb_compose_table_new_from_locale(ctx, locale,
+                                              XKB_COMPOSE_COMPILE_NO_FLAGS);
+        if (!compose_table) {
+            fprintf(stderr, "Couldn't create compose from locale\n");
+            goto out;
+        }
     }
 
-    printf("Compiled successfully from locale %s\n", locale);
+
+    struct xkb_compose_table_iterator *iter = xkb_compose_table_iterator_new(compose_table);
+    while (xkb_compose_table_iterator_next(iter)) {
+        compose_table_print_cb(xkb_compose_table_iterator_current(iter), NULL);
+    }
+    xkb_compose_table_iterator_unref(iter);
```

Comment:
We should this iteration pattern as an example in the docs.

Reply:
Done